helion/runtime/ref_mode.py

```diff
@@ -0,0 +1,109 @@
+from __future__ import annotations
+
+import threading
+from typing import Any
+from typing import Callable
+from typing import Sequence
+
+import torch
+from torch.overrides import BaseTorchFunctionMode
+
+_thread_local = threading.local()
+
+
+def is_ref_mode_enabled() -> bool:
+    """Check if ref mode is currently active."""
+    return getattr(_thread_local, "ref_mode_enabled", False)
+
+
+class RefModeContext:
+    """Context manager to enable ref mode execution."""
+
+    def __enter__(self) -> RefModeContext:
+        self._old_value = getattr(_thread_local, "ref_mode_enabled", False)
+        _thread_local.ref_mode_enabled = True
+        return self
+
+    def __exit__(self, exc_type: object, exc_val: object, exc_tb: object) -> bool:
+        _thread_local.ref_mode_enabled = self._old_value
+        return False
+
+
+class MixedPrecisionMatmulMode(BaseTorchFunctionMode):
+    """Torch function mode that handles mixed precision matmul operations."""
+
+    def __torch_function__(
+        self,
+        func: Callable[..., Any],
+        types: Sequence[type[Any]],
+        args: tuple[Any, ...] = (),
+        kwargs: dict[str, Any] | None = None,
+    ) -> Any:
+        kwargs = kwargs or {}
+
+        if func == torch.addmm:
+            return self._handle_addmm(*args, **kwargs)
+
+        if func == torch.baddbmm:
+            return self._handle_baddbmm(*args, **kwargs)
+
+        # Pass through all other operations
+        return super().__torch_function__(func, types, args, kwargs)
+
+    def _handle_addmm(
+        self,
+        input: torch.Tensor,  # noqa: A002
+        mat1: torch.Tensor,
+        mat2: torch.Tensor,
+        *,
+        beta: float = 1,
+        alpha: float = 1,
+    ) -> torch.Tensor:
+        """Patched addmm that handles dtype mismatches.
+
+        `torch.addmm(fp32, bf16, bf16) -> fp32` is a common usage pattern in Helion kernels,
+        and it throws dtype mismatch error in unpatched torch.addmm op."""
+
+        # Perform matmul in original dtype
+        result = self._handle_matmul(mat1, mat2)
+
+        # Scale the result
+        if alpha != 1:
+            result = result * alpha
+
+        # Add the bias term, converting result to input's dtype if needed
+        if result.dtype != input.dtype:
+            result = result.to(input.dtype)
+
+        if beta == 0:
+            return result
+        return result + (beta * input)
```

Comment:
this is losing precision compared to matmul producing fp32 result, and higher precision is the main reason to use it at all. If helion is using  `torch.addmm(fp32, bf16, bf16)` should it follow pytorch semantics and add out_dtype argument when needed?